[BEAM-7003 BEAM-8639 BEAM-8774] Update Kafka dependencies, enable IT test in Postcommit

piotr-szuberski opened the pull request:
- Update Kafka dependencies to the latest stable versions
- Add tests for previous compatible versions
- Add possibility to use testcontainers in the IT test suite

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

piotr-szuberski commented:
R: @methodmissing Could I ask you for review? I removed the code enforcing kafka  clients 2.1.0 version since it's now 2.6.0 by default. I'm not sure whether I'm not missing something though.

piotr-szuberski commented:
Run Python PreCommit

piotr-szuberski commented:
Run Java KafkaIO Performance Test

methodmissing commented:
I have a branch in progress to run the `io/kafka` tests with more kafka-client versions in https://github.com/apache/beam/compare/master...methodmissing:kafkaio-test-more-versions but haven't had a window to converge on that, but can prioritise otherwise.

Related, but forcing a newer kafka-clients version, I think that'll mostly invalidate `ConsumerSpEL` and `ProducerSpEL` as well.

@lukecwik thoughts on the implications of bumping default kafka-clients up from 1.0.0 to 2.4.1 in 1 fell swoop? It's runtime pluggable, at Shopify we already use 2.4.1 for a long time and moving towards 2.6.0 for Beam but I worry about backwards compatibility. Also admittedly biased because https://github.com/apache/beam/pull/12794 would have been way less complex to implement and users of `1.0.0` in the wild is setting themselves up for other problems moving forward too 🤷‍♂️ :smile: 

cc @sap1ens @vvuibert

lukecwik commented:
> I have a branch in progress to run the `io/kafka` tests with more kafka-client versions in [master...methodmissing:kafkaio-test-more-versions](https://github.com/apache/beam/compare/master...methodmissing:kafkaio-test-more-versions) but haven't had a window to converge on that, but can prioritise otherwise.
> 
> Related, but forcing a newer kafka-clients version, I think that'll mostly invalidate `ConsumerSpEL` and `ProducerSpEL` as well.
> 
> @lukecwik thoughts on the implications of bumping default kafka-clients up from 1.0.0 to 2.4.1 in 1 fell swoop? It's runtime pluggable, at Shopify we already use 2.4.1 for a long time and moving towards 2.6.0 for Beam but I worry about backwards compatibility. Also admittedly biased because #12794 would have been way less complex to implement and users of `1.0.0` in the wild is setting themselves up for other problems moving forward too 🤷‍♂️ 😄
> 
> cc @sap1ens @vvuibert

The purpose of supporting multiple versions is to have greater flexibility for users so the change in #12794 would have been equally complex I believe.

Typically I would lean to allowing this because we typically do perform dependency updates regularly but we have had issues in the past with some libraries like Avro and since we were already doing something special here likely means that it is important to the user community. Its best to ask on dev@beam.apache.org for feedback on this change (@aromanenko-dev and @iemejia should know more about this).

sap1ens commented:
Just two cents about Kafka compatibility, it's been working extremely well since 0.10.2 ([this is a good overview about the design](https://www.confluent.io/blog/upgrading-apache-kafka-clients-just-got-easier/)), so I believe the proposed upgrade here should be very straightforward.

je-ik commented:
I'll just metion one issue I ran into, `poll` can now throw `RebalanceInProgressException` that must be handled: https://kafka.apache.org/26/javadoc/org/apache/kafka/common/errors/RebalanceInProgressException.html.
This applies only to cases when consumer `group id` used, we don't use it by default, but when user specifies it, the handling probably should be added. This change was introduced in KIP https://cwiki.apache.org/confluence/display/KAFKA/KIP-429%3A+Kafka+Consumer+Incremental+Rebalance+Protocol

piotr-szuberski commented:
> This is quite significant change since Beam tended to support most of the Kafka client versions almost from the beginning. I believe that there is still a decent amount of users that uses Beam with quite old Kafka client versions [1]. So, before making a final decision, I propose to do a survey on [user@beam.apache.org](mailto:user@beam.apache.org) and see the results.

I've created a survey on the users mailing list but it was probably not accepted.

aromanenko-dev commented:
@piotr-szuberski What do you mean by "it was not accepted"? I can see you email "[DISCUSS] Update Kafka dependencies in Beam Java SDK" from Oct 22nd on user@

piotr-szuberski commented:
> @piotr-szuberski What do you mean by "it was not accepted"? I can see you email "[DISCUSS] Update Kafka dependencies in Beam Java SDK" from Oct 22nd on user@

I just wasn't subscribed to user@ mailing list before and the thread was disappearing without any message.

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PreCommit

piotr-szuberski commented:
Run Java_Examples_Dataflow PreCommit

piotr-szuberski commented:
@aromanenko-dev @iemejia Kafka previous versions tests (including IT) are now running in Java Postcommit - is this solution acceptable?

aromanenko-dev commented:
@piotr-szuberski Thanks, sorry for delay with answer, I'll take a look asap

aromanenko-dev commented on sdks/java/io/kafka/build.gradle:
```diff
@@ -65,26 +76,68 @@ dependencies {
   testRuntimeOnly library.java.slf4j_jdk14
   testRuntimeOnly project(path: ":runners:direct-java", configuration: "shadow")
-  kafkaVersion210 "org.apache.kafka:kafka-clients:2.1.0"
+  kafkaVersions.each {"kafkaVersion$it.key" "org.apache.kafka:kafka-clients:$it.value"}
```
Will it require to include all these different Kafka client version jars into regular build?

piotr-szuberski replied:
It should do the same what kafkaVersion210 did. I've imported KafkaIO 2.25 jar and I can't see kafka-clients 2.1.0 in the dependencies.
I'm not sure how to import a jar to Gradle locally, I will try it out on Monday.

aromanenko-dev replied:
Could you try to build and publish KafkaIO jar with this command:
```
./gradlew -Ppublishing -PdistMgmtSnapshotsUrl=/path/to/.m2/repository/ -p sdks/java/io/kafka publishToMavenLocal
```
and check out which Kafka clients deps it comtains?

piotr-szuberski replied:
That's super useful! Thank you!
I can't see any Kafka clients in the created jar. Here is the output of :sdks:java:io:kafka:dependencies

```
compileClasspath - Compile classpath for source set 'main'.
\--- org.apache.beam:beam-sdks-java-io-kafka:2.27.0-SNAPSHOT
     +--- org.checkerframework:checker-qual:3.7.0 FAILED
     +--- org.apache.beam:beam-vendor-guava-26_0-jre:0.1
     +--- org.apache.beam:beam-sdks-java-core:2.27.0-SNAPSHOT FAILED
     +--- org.apache.beam:beam-sdks-java-expansion-service:2.27.0-SNAPSHOT FAILED
     +--- org.slf4j:slf4j-api:1.7.30
     +--- joda-time:joda-time:2.10.5
     +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
     +--- com.fasterxml.jackson.core:jackson-databind:2.10.2
     |    +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
     |    \--- com.fasterxml.jackson.core:jackson-core:2.10.2
     +--- org.springframework:spring-expression:4.3.18.RELEASE FAILED
     \--- io.confluent:kafka-avro-serializer:5.3.2
          +--- org.apache.avro:avro:1.8.1
          |    +--- org.codehaus.jackson:jackson-core-asl:1.9.13
          |    +--- org.codehaus.jackson:jackson-mapper-asl:1.9.13
          |    |    \--- org.codehaus.jackson:jackson-core-asl:1.9.13
          |    +--- com.thoughtworks.paranamer:paranamer:2.7
          |    +--- org.xerial.snappy:snappy-java:1.1.1.3
          |    +--- org.apache.commons:commons-compress:1.8.1
          |    +--- org.tukaani:xz:1.5
          |    \--- org.slf4j:slf4j-api:1.7.7 -> 1.7.30
          +--- io.confluent:kafka-schema-registry-client:5.3.2
          |    +--- io.confluent:common-config:5.3.2
          |    |    +--- io.confluent:common-utils:5.3.2
          |    |    |    +--- org.slf4j:slf4j-api:1.7.26 -> 1.7.30
          |    |    |    \--- com.101tec:zkclient:0.10
          |    |    |         \--- org.slf4j:slf4j-api:1.6.1 -> 1.7.30
          |    |    \--- org.slf4j:slf4j-api:1.7.26 -> 1.7.30
          |    +--- org.apache.avro:avro:1.8.1 (*)
          |    +--- com.fasterxml.jackson.core:jackson-databind:2.9.10.1 -> 2.10.2 (*)
          |    \--- io.confluent:common-utils:5.3.2 (*)
          +--- io.confluent:common-config:5.3.2 (*)
          \--- io.confluent:common-utils:5.3.2 (*)

compileOnly - Compile only dependencies for source set 'main'. (n)
No dependencies

default - Configuration for default artifacts. (n)
No dependencies

implementation - Implementation only dependencies for source set 'main'. (n)
No dependencies

runtimeClasspath - Runtime classpath of source set 'main'.
\--- org.apache.beam:beam-sdks-java-io-kafka:2.27.0-SNAPSHOT
     +--- org.checkerframework:checker-qual:3.7.0 FAILED
     +--- org.apache.beam:beam-vendor-guava-26_0-jre:0.1
     +--- org.apache.beam:beam-sdks-java-core:2.27.0-SNAPSHOT FAILED
     +--- org.apache.beam:beam-sdks-java-expansion-service:2.27.0-SNAPSHOT FAILED
     +--- org.slf4j:slf4j-api:1.7.30
     +--- joda-time:joda-time:2.10.5
     +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
     +--- com.fasterxml.jackson.core:jackson-databind:2.10.2
     |    +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
     |    \--- com.fasterxml.jackson.core:jackson-core:2.10.2
     +--- org.springframework:spring-expression:4.3.18.RELEASE FAILED
     \--- io.confluent:kafka-avro-serializer:5.3.2
          +--- org.apache.avro:avro:1.8.1
          |    +--- org.codehaus.jackson:jackson-core-asl:1.9.13
          |    +--- org.codehaus.jackson:jackson-mapper-asl:1.9.13
          |    |    \--- org.codehaus.jackson:jackson-core-asl:1.9.13
          |    +--- com.thoughtworks.paranamer:paranamer:2.7
          |    +--- org.xerial.snappy:snappy-java:1.1.1.3
          |    +--- org.apache.commons:commons-compress:1.8.1
          |    +--- org.tukaani:xz:1.5
          |    \--- org.slf4j:slf4j-api:1.7.7 -> 1.7.30
          +--- io.confluent:kafka-schema-registry-client:5.3.2
          |    +--- io.confluent:common-config:5.3.2
          |    |    +--- io.confluent:common-utils:5.3.2
          |    |    |    +--- org.slf4j:slf4j-api:1.7.26 -> 1.7.30
          |    |    |    \--- com.101tec:zkclient:0.10
          |    |    |         \--- org.slf4j:slf4j-api:1.6.1 -> 1.7.30
          |    |    \--- org.slf4j:slf4j-api:1.7.26 -> 1.7.30
          |    +--- org.apache.avro:avro:1.8.1 (*)
          |    +--- com.fasterxml.jackson.core:jackson-databind:2.9.10.1 -> 2.10.2 (*)
          |    \--- io.confluent:common-utils:5.3.2 (*)
          +--- io.confluent:common-config:5.3.2 (*)
          \--- io.confluent:common-utils:5.3.2 (*)

runtimeElements - Elements of runtime for main. (n)
No dependencies

runtimeOnly - Runtime only dependencies for source set 'main'. (n)
No dependencies

testAnnotationProcessor - Annotation processors and their dependencies for source set 'test'.
No dependencies

testCompileClasspath - Compile classpath for source set 'test'.
\--- org.apache.beam:beam-sdks-java-io-kafka:2.27.0-SNAPSHOT
     +--- org.checkerframework:checker-qual:3.7.0 FAILED
     +--- org.apache.beam:beam-vendor-guava-26_0-jre:0.1
     +--- org.apache.beam:beam-sdks-java-core:2.27.0-SNAPSHOT FAILED
     +--- org.apache.beam:beam-sdks-java-expansion-service:2.27.0-SNAPSHOT FAILED
     +--- org.slf4j:slf4j-api:1.7.30
     +--- joda-time:joda-time:2.10.5
     +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
     +--- com.fasterxml.jackson.core:jackson-databind:2.10.2
     |    +--- com.fasterxml.jackson.core:jackson-annotations:2.10.2
     |    \--- com.fasterxml.jackson.core:jackson-core:2.10.2
     +--- org.springframework:spring-expression:4.3.18.RELEASE FAILED
     \--- io.confluent:kafka-avro-serializer:5.3.2
          +--- org.apache.avro:avro:1.8.1
          |    +--- org.codehaus.jackson:jackson-core-asl:1.9.13
          |    +--- org.codehaus.jackson:jackson-mapper-asl:1.9.13
          |    |    \--- org.codehaus.jackson:jackson-core-asl:1.9.13
          |    +--- com.thoughtworks.paranamer:paranamer:2.7
          |    +--- org.xerial.snappy:snappy-java:1.1.1.3
          |    +--- org.apache.commons:commons-compress:1.8.1
          |    +--- org.tukaani:xz:1.5
          |    \--- org.slf4j:slf4j-api:1.7.7 -> 1.7.30
          +--- io.confluent:kafka-schema-registry-client:5.3.2
          |    +--- io.confluent:common-config:5.3.2
          |    |    +--- io.confluent:common-utils:5.3.2
          |    |    |    +--- org.slf4j:slf4j-api:1.7.26 -> 1.7.30
          |    |    |    \--- com.101tec:zkclient:0.10
          |    |    |         \--- org.slf4j:slf4j-api:1.6.1 -> 1.7.30
          |    |    \--- org.slf4j:slf4j-api:1.7.26 -> 1.7.30
          |    +--- org.apache.avro:avro:1.8.1 (*)
          |    +--- com.fasterxml.jackson.core:jackson-databind:2.9.10.1 -> 2.10.2 (*)
          |    \--- io.confluent:common-utils:5.3.2 (*)
          +--- io.confluent:common-config:5.3.2 (*)
          \--- io.confluent:common-utils:5.3.2 (*)
```

piotr-szuberski replied:
Sorry, I missed the request for logging the Kafka version. I'm working on it now.

piotr-szuberski replied:
Done.

aromanenko-dev commented on sdks/java/io/kafka/build.gradle:
```diff
@@ -65,26 +76,68 @@ dependencies {
   }
 }
 
-task kafkaVersion210Test(type: Test) {
```
I think this task was used to run KafkaIO unit tests against specific client version. It's not a case now?

piotr-szuberski replied:
This task is still there at
```
kafkaVersions.each {kv ->
  task "kafkaVersion${kv.key}Test"(type: Test)
  [...]
```
Though I used 2.1.1 version - should we leave 2.1.0?

aromanenko-dev replied:
I think it should fine since 2.1.0 was used as a minimum version where headers were added (if I'm not mistaken).

piotr-szuberski replied:
I'm not sure as well but I guess so. I'm having a difficulty with the recently added Kafka streaming performance tests so  I'm changing this PR to WIP.

If I understand correctly these streaming tests can't be run on DirectRunner? They never stop even with `readTimeout` set to 1min

piotr-szuberski replied:
Kafka performance tests have been failing (timeout) since https://github.com/apache/beam/pull/13282 merge.

I guess this PR should wait until it's fixed?

https://ci-beam.apache.org/job/beam_PerformanceTests_Kafka_IO/

aromanenko-dev replied:
@boyuanzz Could you take a look on https://ci-beam.apache.org/job/beam_PerformanceTests_Kafka_IO/1563/ ? After rebase with latest KafkaIOIT changes, It fails because of timeout.

boyuanzz replied:
Looking.

boyuanzz replied:
It seems like the sdk docker image is cleaned up incorrectly before test finishes. I can have a quick fix for that.

boyuanzz replied:
So the problem is introduced by https://github.com/apache/beam/pull/13283/files

aromanenko-dev replied:
@boyuanzz Thanks! I guess it's should fixed now (#13356 ), right?

aromanenko-dev reviewed (commented):
Thanks! I added a couple of comments, ptal.

Also, I'd suggest to add, at least, into logs the Kafka client version information for test results that run against specific version, like here:
https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/490/testReport/org.apache.beam.sdk.io.kafka/KafkaIOIT/

And this PR will require rebase because of #13282

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java PreCommit

piotr-szuberski commented:
Run Java_Examples_Dataflow PreCommit

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java KafkaIO Performance Test

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PreCommit

piotr-szuberski commented:
For some reason the printed versions are correct when run locally and on Jenkins it's always `INFO: Running test for Kafka version 0.11.0.3.`. I'll add some better asserion to IT tests and try to figure out what's going on. Maybe Gradle runs everything in parallel and overwrites the previous kafka clients jar versions?

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
@aromanenko-dev Ok, despite printing all the time the same version it looks like it's working. I tried to put Kafka 0.9 among the Kafka versions and it indeed failed (still printing 0.11.3).
So I removed the version printing and added a new test to the IT that verifies that the version from AppInfoParser is the same as in the kafka versions map.

piotr-szuberski commented:
Run Java_Examples_Dataflow PreCommit

aromanenko-dev commented:
@piotr-szuberski Actually, I proposed to print out the Kafka client version because now [here](https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/507/testReport/org.apache.beam.sdk.io.kafka/KafkaIOIT/) we can see the list of the executed tests but if some of them failed, we can't say for sure which Kafka version were used. Do I miss something?

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
> @piotr-szuberski Actually, I proposed to print out the Kafka client version because now [here](https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/507/testReport/org.apache.beam.sdk.io.kafka/KafkaIOIT/) we can see the list of the executed tests but if some of them failed, we can't say for sure which Kafka version were used. Do I miss something?

You're right about it, I've tried to print some logs from the tests but they are not visible on Jenkins.

I'll try to add some gradle-level error logs. It looks like printing "info" logs is not sufficient because postcommit task doesn't have '--info' switch.

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
@aromanenko-dev Ok I'm out of ideas.
Jenkins takes `**/build/test-results/**.xml` files to print the result of executed tests.
All of the kafka versions tests are the same (KafkaIOIT, KafkaIOTest), so they override each other in the test result dashboard.
I tried to rename these xml files but it looks like Jenkins cares about the info inside them.
So there seems to be no simple way to just print the version for each test execution.
Maybe it's sufficient to just know that some Kafka version is broken and run the tests locally? They don't take much time and everything is printed nicely when run on a local machine.

aromanenko-dev commented:
@piotr-szuberski It seems a bit odd but I don't think it's a blocker for this PR. In any case, we have a Kafka version in the gradle task name. So if it fails then the corresponding task will fail too, right? And we can see that in Jenkins job logs, like [here](https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/511/consoleText). 

So, in general it LGTM. Do you have something to add to this PR? Otherwise, we can merge that I think.

aromanenko-dev commented:
@chamikaramj @boyuanzz I guess you can be interested in this PR too.

piotr-szuberski commented:
Run Java PostCommit

piotr-szuberski commented:
> @piotr-szuberski It seems a bit odd but I don't think it's a blocker for this PR. In any case, we have a Kafka version in the gradle task name. So if it fails then the corresponding task will fail too, right? And we can see that in Jenkins job logs, like [here](https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/511/consoleText).

Yes, exactly like [here](https://ci-beam.apache.org/job/beam_PostCommit_Java_PR/506/console):
```
00:18:25 Execution failed for task ':sdks:java:io:kafka:kafkaVersion090BatchIT'.
```
> 
> So, in general it LGTM. Do you have something to add to this PR? Otherwise, please undraft it to make it possible to merge.

Great! I've made some cleanup and removed the xml renaming part. When it's green then it's ready for merge

piotr-szuberski commented:
Run Java_Examples_Dataflow PreCommit

piotr-szuberski commented:
@aromanenko-dev From my side it's ready to merge.

aromanenko-dev reviewed (approved):
LGTM, thanks for your work, @piotr-szuberski !

ibzib commented:
@piotr-szuberski @aromanenko-dev @iemejia 

This broke the Spark postcommit: BEAM-11375

It looks like the cause is the mismatch between `kafka_2.11` (2.4.1) and `kafka-clients` (2.6.0). Kafka 2.5.0 deprecated support for Scala 2.11, so I'm guessing we can't bring it up to date until upgrading Spark to Scala 2.12?

piotr-szuberski commented:
> @piotr-szuberski @aromanenko-dev @iemejia
> 
> This broke the Spark postcommit: [BEAM-11375](https://issues.apache.org/jira/browse/BEAM-11375)
> 
> It looks like the cause is the mismatch between `kafka_2.11` (2.4.1) and `kafka-clients` (2.6.0). Kafka 2.5.0 deprecated support for Scala 2.11, so I'm guessing we can't bring it up to date until upgrading Spark to Scala 2.12?

Sorry for that, I didn't take it into account. I'll try to either update the scala version of Kafka dependency or downgrade kafka clients to 2.4.1
